core/chains/evm/client/simulated_backend_client.go

```diff
@@ -66,20 +66,35 @@ var (
 // SimulatedBackendClient is an Client implementation using a simulated
 // blockchain backend. Note that not all RPC methods are implemented here.
 type SimulatedBackendClient struct {
-	b       *backends.SimulatedBackend
-	t       testing.TB
-	chainId *big.Int
+	b                    evmtypes.Backend // *simulated.Backend, or something satisfying same interface
+	client               simulated.Client
+	t                    testing.TB
+	chainId              *big.Int
+	optimismMode         bool
+	headByNumberCallback func(ctx context.Context, c *SimulatedBackendClient, n *big.Int) error
 }
 
 // NewSimulatedBackendClient creates an eth client backed by a simulated backend.
-func NewSimulatedBackendClient(t testing.TB, b *backends.SimulatedBackend, chainId *big.Int) *SimulatedBackendClient {
+func NewSimulatedBackendClient(t testing.TB, b *simulated.Backend, chainId *big.Int) *SimulatedBackendClient {
 	return &SimulatedBackendClient{
 		b:       b,
+		client:  b.Client(),
 		t:       t,
 		chainId: chainId,
 	}
 }
 
+// Switch to a new backend client (simulating an rpc failover event)
+// If optimismMode = true, the new backend will exhibit the non-geth behavior of optimism (and some other rpc clients),
+// where success rather than an error code is returned when a call to FilterLogs() fails to find the block hash
+// requested. This combined with a failover event can lead to the "eventual consistency" behavior that Backup LogPoller
+// and other solutions were designed to recover from.
+func (c *SimulatedBackendClient) SetBackend(backend evmtypes.Backend, optimismMode bool) {
+	c.optimismMode = optimismMode
```

Comment:
Would it make sense to pass ChainType here, so the API stays a bit more forward compatible, instead of being so closely tied to optimism? Can even leave the internals as-is for now:
```suggestion
func (c *SimulatedBackendClient) SetBackend(backend evmtypes.Backend, chainType evm.ChainType) {
	c.optimismMode = chainType == evm.ChainTypeOptimism
```